### PR TITLE
[feature]: initial mso backup schedule

### PR DESCRIPTION
**Pull request by akinross**
Note:

Added backup scheduler functionality

Issue: #85

### CODE REVIEW

File: tests/integration/targets/mso_backup_schedule/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    mso_info: &mso_info
+      host: '{{ mso_hostname }}'
+      username: '{{ mso_username }}'
+      password: '{{ mso_password }}'
+      validate_certs: '{{ mso_validate_certs | default(false) }}'
+      use_ssl: '{{ mso_use_ssl | default(true) }}'
+      use_proxy: '{{ mso_use_proxy | default(true) }}'
+      output_level: '{{ mso_output_level | default("info") }}'
+
+- name: Query MSO version
```
Comment (anvitha-jain) [inline tests/integration/targets/mso_backup_schedule/tasks/main.yml:24]: we do not need this query if it is not used later in this file.

**Review by anvitha-jain — CHANGES_REQUESTED** [review]

File: tests/integration/targets/mso_backup_schedule/tasks/main.yml (hunk 1 above)
Comment (akinross) [reply tests/integration/targets/mso_backup_schedule/tasks/main.yml]: ok removed

File: plugins/modules/mso_backup_schedule.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    - The interval unit type
+    choices: [ hours, days ]
+    type: str
+  interval_length:
+    description:
+    - Amount of hours or days for schedule to trigger
+    type: int
+  remote_location:
+    description:
+    - The remote location's name for the backup to be stored
```
Comment (lhercot) [inline plugins/modules/mso_backup_schedule.py:47]: The remote location's name where the backup should be stored

File: plugins/modules/mso_backup_schedule.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    description:
+    - The date to start the scheduler in format YYYY-MM-DD
+    - If no date is provided, the current date will be used.
+    type: str
+  start_time:
+    description:
+    - The time to start the scheduler in format HH:MM:SS
+    - If no time is provided, midnight "00:00:00" will be used.
+    type: str
+  interval_unit:
```
Comment (lhercot) [inline plugins/modules/mso_backup_schedule.py:36]: UI uses frequency so I think we should use frequency_unit

File: plugins/modules/mso_backup_schedule.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    description:
+    - The interval unit type
+    choices: [ hours, days ]
+    type: str
+  interval_length:
```
Comment (lhercot) [inline plugins/modules/mso_backup_schedule.py:41]: UI uses frequency so I think we should use frequency_length

File: plugins/modules/mso_backup_schedule.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    description:
+    - Amount of hours or days for schedule to trigger
```
Comment (lhercot) [inline plugins/modules/mso_backup_schedule.py:43]: for the schedule trigger frequency

File: plugins/modules/mso_backup_schedule.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    - Use C(present) or C(absent) for adding or removing.
+    - Use C(query) for listing an object or multiple objects.
+    type: str
+    choices: [ absent, present, query ]
+    default: present
+extends_documentation_fragment: cisco.mso.modules
+'''
+
+EXAMPLES = r'''
+- name: Get set backup schedule
```
Comment (lhercot) [inline plugins/modules/mso_backup_schedule.py:66]: Get current backup schedule

File: plugins/modules/mso_backup_schedule.py [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+            mso.existing = mso.request(api_path, method='DELETE')
+
+    elif state == 'present':
+        mso.previous = mso.existing
+
+        remote_location_info = mso.lookup_remote_location(remote_location)
+
+        # If no time has been provided default to midnight "00:00:00".
+        try:
+            start_time = time.fromisoformat(start_time) if start_time else time()
```
Comment (lhercot) [inline plugins/modules/mso_backup_schedule.py:162]: is time() 00:00:00 or the current time?

File: plugins/modules/mso_backup_schedule.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+        except Exception as e:
+            module.fail_json(msg="Failed to parse time format 'HH:MM:SS' %s', %s" % (start_time, e))
```
Comment (lhercot) [inline plugins/modules/mso_backup_schedule.py:164]: why is there a single quote only after the first %s ?
Would probably use "Failed to parse %s with time format 'HH:MM:SS', %s"

File: plugins/modules/mso_backup_schedule.py [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+
+        # If no date has been provided default to current date.
+        try:
+            start_date = date.fromisoformat(start_date) if start_date else date.today()
+        except Exception as e:
+            module.fail_json(msg="Failed to parse date format 'YYYY-MM-DD' '%s', %s" % (start_date, e))
```
Comment (lhercot) [inline plugins/modules/mso_backup_schedule.py:170]: "Failed to parse '%s' with date format 'YYYY-MM-DD', %s"

**Review by lhercot — CHANGES_REQUESTED** [review]

File: plugins/modules/mso_backup_schedule.py (hunk 3 above)
Comment (akinross) [reply plugins/modules/mso_backup_schedule.py:36]: ok

File: plugins/modules/mso_backup_schedule.py (hunk 4 above)
Comment (akinross) [reply plugins/modules/mso_backup_schedule.py:41]: ok

File: plugins/modules/mso_backup_schedule.py (hunk 5 above)
Comment (akinross) [reply plugins/modules/mso_backup_schedule.py:43]: ok

File: plugins/modules/mso_backup_schedule.py (hunk 6 above)
Comment (akinross) [reply plugins/modules/mso_backup_schedule.py:66]: ok

File: plugins/modules/mso_backup_schedule.py (hunk 7 above)
Comment (akinross) [reply plugins/modules/mso_backup_schedule.py:162]: 00:00:00

File: plugins/modules/mso_backup_schedule.py (hunk 8 above)
Comment (akinross) [reply plugins/modules/mso_backup_schedule.py:164]: typo, have changed

File: plugins/modules/mso_backup_schedule.py (hunk 9 above)
Comment (akinross) [reply plugins/modules/mso_backup_schedule.py:170]: typo, have changed

File: plugins/modules/mso_backup_schedule.py (hunk 2 above)
Comment (akinross) [reply plugins/modules/mso_backup_schedule.py:47]: Copied this from mso_backup module to keep consistent, shall I change there as well?

https://github.com/akinross/ansible-mso/blob/backup-scheduler/plugins/modules/mso_backup.py#L36

Comment (lhercot) [reply plugins/modules/mso_backup_schedule.py:47]: Let's fix both them.

**Review by lhercot — APPROVED** [review]
LGTM

**Pull request merged**